modules/internal/DefaultRectangular.chpl

```diff
@@ -842,24 +841,31 @@ module DefaultRectangular {
           sum += (ind(i) - off(i)) * blk(i) / abs(str(i)):idxType;
         return sum;
       } else {
-        var sum = if earlyShiftData then 0:idxType else origin;
-
-        // If the user asserts that there is no slicing in their program,
-        // then blk(rank) == 1. Knowing this, we need not multiply the final
-        // ind(...) by anything. This may lead to performance improvements for
-        // array accesses.
-        if assertNoSlicing {
-          for param i in 1..rank-1 {
-            sum += ind(i) * blk(i);
+        if (rank == 1 && earlyShiftData) {
+          if __primitive("optimize_array_blk_mult") {
+            return ind(1);
+          } else {
+            return ind(1) * blk(1);
```

Comment:
good point, I'll add a comment. And actually, that was with dernom turned on.


We can do the same sort of thing for multidim, but the generated code for that is already much uglier and it's hard to notice the improvement


```chapel
  for i in 1..10 do A[i, i] = i;
```

```c
  for (i = INT64(1); ((i <= _ic__F1_high)); i += INT64(1)) {
    sum = INT64(0);
    sum += ((int64_t)((i * coerce_tmp)));
    sum += i;
    call_tmp8 = (coerce_tmp2 + sum);
    *(call_tmp8) = i;
  }
```

vs

```c
  for (i = INT64(1); ((i <= _ic__F1_high)); i += INT64(1)) {
    sum = ((int64_t)((i * coerce_tmp)));
    sum += i;
    call_tmp8 = (coerce_tmp2 + sum);
    *(call_tmp8) = i;
  }
```